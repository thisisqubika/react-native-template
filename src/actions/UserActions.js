import { UserController } from '@/controllers';
import { strings } from '@/localization';

export const TYPES = {
  CLEAR_STORE: 'CLEAR_STORE',
  LOGIN: 'LOGIN',
  LOGIN_REQUEST: 'LOGIN_REQUEST',
  LOGIN_ERROR: 'LOGIN_ERROR',
  LOGIN_SUCCESS: 'LOGIN_SUCCESS',
};

const loginRequest = () => ({
  type: TYPES.LOGIN_REQUEST,
  payload: null,
});

const loginSuccess = (user) => ({
  type: TYPES.LOGIN_SUCCESS,
  payload: { user },
});

const loginError = (error) => ({
  type: TYPES.LOGIN_ERROR,
  payload: { error },
});

const clearStore = () => ({
  type: TYPES.CLEAR_STORE,
  payload: null,
});

export const login =
  (username, password) =>
  async (dispatch, _, { demoMode, networkService }) => {
    try {
      dispatch(loginRequest());
      const userController = new UserController(networkService);
      const { data } = await userController.login({ username, password, demoMode });
      if (!demoMode) {
        networkService.setAccessToken(data.user.accessToken);
      }
      dispatch(loginSuccess(data.user));
    } catch ({ data }) {
      dispatch(loginError(data?.error ?? strings.login.invalidCredentials));
    }
  };

export const logout =
  () =>
  async (dispatch, _, { demoMode, networkService }) => {
    try {
      const userController = new UserController(networkService);
      await userController.logout({ demoMode });
    } finally {
      networkService.clearAccessToken();
      dispatch(clearStore());
    }
  };
