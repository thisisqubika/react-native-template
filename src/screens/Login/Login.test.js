import { fireEvent, render, waitForElementToBeRemoved } from '@testing-library/react-native';
import React from 'react';
import * as UserActions from '@/actions/UserActions';
import { strings } from '@/localization';
import { mockLoginNetworkService } from '@/mocks';
import { Login } from '@/screens/Login/Login';
import { withProviders } from '@/test-utils';

describe('Login', () => {
  it('should match the snapshot', () => {
    const { toJSON } = render(
      withProviders(<Login />, { networkService: mockLoginNetworkService })
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should make the request with valid credentials', async () => {
    const loginSpy = jest.spyOn(UserActions, 'login');

    const { getByHintText, getByText } = render(
      withProviders(<Login />, { networkService: mockLoginNetworkService })
    );

    const submitButton = getByText(strings.login.button);
    const usernameInput = getByHintText(strings.login.usernameHint);
    const passwordInput = getByHintText(strings.login.passwordHint);

    fireEvent.changeText(usernameInput, 'username');
    fireEvent.changeText(passwordInput, 'password');
    fireEvent.press(submitButton);

    expect(loginSpy).toHaveBeenCalledWith('username', 'password');

    await waitForElementToBeRemoved(() => getByText(strings.common.loading));

    loginSpy.mockRestore();
  });

  it('should show an error with invalid credentials', () => {
    const { getByHintText, getByText } = render(
      withProviders(<Login />, { networkService: mockLoginNetworkService })
    );

    const submitButton = getByText(strings.login.button);
    const usernameInput = getByHintText(strings.login.usernameHint);
    const passwordInput = getByHintText(strings.login.passwordHint);

    fireEvent.changeText(usernameInput, 'username');
    fireEvent.changeText(passwordInput, 'invalidPassword');
    fireEvent.press(submitButton);

    expect(getByText(strings.login.invalidCredentials)).toBeTruthy();
  });
});
