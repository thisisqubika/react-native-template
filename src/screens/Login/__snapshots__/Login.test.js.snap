// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 40,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 5,
          "padding": 16,
          "width": "100%",
        },
        Object {
          "elevation": 5,
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.25,
          "shadowRadius": 3.84,
        },
        Object {
          "backgroundColor": "rgb(0, 122, 255)",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "borderBottomWidth": 0.5,
          "marginBottom": 16,
          "paddingBottom": 8,
          "paddingTop": 24,
        }
      }
    >
      <Text
        animatedStyle={
          Object {
            "value": Object {
              "opacity": 1,
              "transform": Array [
                Object {
                  "scale": 1,
                },
                Object {
                  "translateX": -0,
                },
                Object {
                  "translateY": 24,
                },
              ],
            },
          }
        }
        collapsable={false}
        onLayout={[Function]}
        style={
          Object {
            "color": "gray",
            "left": 8,
            "opacity": 1,
            "position": "absolute",
            "top": 0,
            "transform": Array [
              Object {
                "scale": 1,
              },
              Object {
                "translateX": -0,
              },
              Object {
                "translateY": 24,
              },
            ],
          }
        }
      >
        Username
      </Text>
      <TextInput
        accessibilityHint="Enter your username"
        accessibilityLabel="Username"
        autoCapitalize="none"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        style={
          Object {
            "paddingHorizontal": 8,
          }
        }
        value=""
      />
    </View>
    <View
      style={
        Object {
          "borderBottomWidth": 0.5,
          "marginBottom": 16,
          "paddingBottom": 8,
          "paddingTop": 24,
        }
      }
    >
      <Text
        animatedStyle={
          Object {
            "value": Object {
              "opacity": 1,
              "transform": Array [
                Object {
                  "scale": 1,
                },
                Object {
                  "translateX": -0,
                },
                Object {
                  "translateY": 24,
                },
              ],
            },
          }
        }
        collapsable={false}
        onLayout={[Function]}
        style={
          Object {
            "color": "gray",
            "left": 8,
            "opacity": 1,
            "position": "absolute",
            "top": 0,
            "transform": Array [
              Object {
                "scale": 1,
              },
              Object {
                "translateX": -0,
              },
              Object {
                "translateY": 24,
              },
            ],
          }
        }
      >
        Password
      </Text>
      <TextInput
        accessibilityHint="Enter your password"
        accessibilityLabel="Password"
        autoCapitalize="none"
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        secureTextEntry={true}
        style={
          Object {
            "paddingHorizontal": 8,
          }
        }
        textContentType="password"
        value=""
      />
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderColor": "rgb(216, 216, 216)",
          "borderRadius": 5,
          "borderWidth": 1,
          "justifyContent": "center",
          "marginTop": 24,
          "opacity": 1,
          "padding": 10,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "rgb(28, 28, 30)",
            },
            Object {
              "fontSize": 16,
              "fontWeight": "700",
            },
            null,
          ]
        }
      >
        Login
      </Text>
    </View>
  </View>
</View>
`;
